Add section in event filter reference for JS functions and add new SQE to SQE reference

hillaryfraley opened the pull request:
## Description
Adds section to cover three new JS functions in event filter reference:
- sensu.EventStatus
- sensu.ListEvents
- sensu.FetchEvent

Adds sensu.CheckDependencies SQE to the SQE reference.

Adds JS functions to list of commercial features.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2632

hillaryfraley committed:
Add section in event filter reference for JS functions

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,97 @@ spec:
+
+The `sensu.CheckDependencies` function takes zero or more checks as arguments.
+It returns `true` if all the specified checks are passing or `false` if any of the specified checks are failing.
+If you do not specify any checks, the function returns `true`.
```
Is there a reason to specify no checks? What would that tell you?

echlebek replied:
No, it's just the base case of the function.

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,97 @@ spec:
+
+### `sensu.EventStatus`
+
+The `sensu.EventStatus` function takes zero or more checks as arguments.
```
Is it correct that this takes checks as arguments?

echlebek replied:
See the examples for the spec of what the function takes. It can either take string check names, or objects that specify an entity name and check name. If only the check name is given, then the entity is computed from the event that is being filtered.

hillaryfraley replied:
I figured that but the gist just said it takes the same arguments as sensu.CheckDependencies so I wanted to make sure.

hillaryfraley committed:
Remove duplicate sentence

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,95 @@ spec:
+
+### `sensu.ListEvents`
+
+The `sensu.ListEvents` function returns an array of all events in the specified namespace.
```
What if you do not specify a namespace?

echlebek replied:
Sorry, that was poorly worded. The namespace cannot be supplied by the user. It is automatically determined by the event being filtered.

hillaryfraley committed:
Incorporate Eric's clarifications

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+{{< code javascript >}}
+sensu.CheckDependencies("database", "disk")
+{{< /code >}}
+
```
I'm thinking it would be better to provide these examples in an actual filter expression...

echlebek replied:
So, to be clear, `sensu.CheckDependencies("database", "disk")` is a full filter expression on its own.

echlebek replied:
The others are more complex. We don't intend for most people to use them, and are really only provided for tinkering.

hillaryfraley replied:
I'm sorry, I wasn't clear and also may not have been thinking about this clearly. I was wondering if we shouldn't provide a complete example filter definition that includes one of these functions as the expression.

hillaryfraley replied:
I added an example above of what I was thinking. Does this make sense at all? Or is it just more confusing? edit: also I'm not sure my example is correct in any case.

echlebek replied:
Sorry, my earlier comment was wrong.

No, this does not make sense. Sensu Query Expressions have to return a boolean. `sensu.FetchEvent` returns a complex data structure (an event).

You'd have to do something like
```
"expressions": [
  '(function () { var diskEvent = sensu.FetchEvent("router01", "disk"); if (diskEvent == nil) { return false; } return diskEvent.check.status == 0; })()'
]
```

echlebek replied:
I don't believe these functions belong inlined in the Sensu Query Expressions docs. They are building blocks for creating Sensu Query Expressions (except for sensu.CheckDependencies, which is a Sensu Query Expression, because it returns a boolean)

hillaryfraley replied:
^ I have them in the event filter reference rather than the SQE reference. Is that what you mean?

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+
+### `sensu.FetchEvent`
+
+The `sensu.FetchEvent` function loads the Sensu event that corresponds to the specified entity and check names.
```
```suggestion
The `sensu.FetchEvent` function loads the Sensu events that correspond to the specified entity and check names.
```

hillaryfraley committed:
Update content/sensu-go/6.0/reference/filters.md

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+- `sensu.CheckStatus`
+- `sensu.ListEvents`
+
+You can use these functions to query your event stores.
```
```suggestion
You can use these functions to query your event stores.
For example, the following event filter allows handling only for events for the `server01` entity *and* the `disk` check:

{{< language-toggle >}}

{{< code yml >}}
type: EventFilter
api_version: core/v2
metadata:
  name: server01_disk_events
  namespace: default
spec:
  action: allow
  expressions:
  - sensu.FetchEvent("server01", "disk")
{{< /code >}}

{{< code json >}}
{
  "type": "EventFilter",
  "api_version": "core/v2",
  "metadata": {
    "name": "server01_disk_events",
    "namespace": "default"
  },
  "spec": {
    "action": "allow",
    "expressions": [
      sensu.FetchEvent("server01", "disk")
    ]
  }
}
{{< /code >}}

{{< /language-toggle >}}
```

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+sensu.FetchEvent("server01", "disk")
+{{< /code >}}
+
+You can only load events from the same namespace as the event being queried.
```
Something seems not quite right with this statement. Should it be "the entity being queried" instead? As written, it seems circular maybe (...events from...the event being queried).

echlebek replied:
That should say "the event being filtered"

hillaryfraley replied:
ohhh right I see heh. I was trying to avoid using "filtered" in a general sense but we are actually talking about the subject event for the event filter. I get it now, thank you.

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+sensu.FetchEvent("server01", "disk")
+{{< /code >}}
+
+You can only load events from the same namespace as the event being queried.
```
```suggestion
You can only load events from the same namespace as the event being filtered.
```

hillaryfraley commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -832,6 +832,96 @@ spec:
+
+### `sensu.ListEvents`
+
+The `sensu.ListEvents` function returns an array of all events in the same namespace as the event being queried.
```
```suggestion
The `sensu.ListEvents` function returns an array of all events in the same namespace as the event being filtered.
```

hillaryfraley committed:
Correct "queried" to "filtered"

hillaryfraley committed:
Merge branch 'master' into filter-js-functions

hillaryfraley committed:
Update with example from Eric and move section to a better spot

hillaryfraley commented:
@echlebek I added your example, moved the section up a bit to a more logical spot in the doc, and added a link to this section in our list of commercial features. Wdyt?

edit: I also moved sensu.CheckDependencies to the SQE reference.

hillaryfraley committed:
Move sensu.CheckDependencies to SQE reference

echlebek commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -366,6 +366,83 @@ For more information about event attributes, see the [event reference][28].
+
+### `sensu.FetchEvent`
+
+The `sensu.FetchEvent` function loads the Sensu events that correspond to the specified entity and check names.
```
`FetchEvent` loads only a single event.

hillaryfraley replied:
```suggestion
The `sensu.FetchEvent` function loads the Sensu event that corresponds to the specified entity and check names.
```

hillaryfraley committed:
Update to reflect sensu.FetchEvent returns a single event.

nikkictl commented on content/sensu-go/6.0/reference/filters.md:
```diff
@@ -366,6 +366,83 @@ For more information about event attributes, see the [event reference][28].
+
+{{< code javascript >}}
+"expressions": [
+  '(function () { var diskEvent = sensu.FetchEvent("router01", "disk"); if (diskEvent == nil) { return false; } return diskEvent.check.status == 0; })()'
```
You used `server01` as the entity name in the example above, might be better if we stay consistent.

hillaryfraley committed:
Incorporate Nikki's suggestion

nikkictl reviewed (approved):
Not sure if there are other outstanding comments, but this LGTM!

hillaryfraley committed:
Merge branch 'master' into filter-js-functions

hillaryfraley committed:
Merge branch 'master' into filter-js-functions
